aten/src/ATen/native/xnnpack/Convolution.cpp

```diff
@@ -0,0 +1,309 @@
+      MemoryFormat::ChannelsLast);
+
+  const xnn_status setup_status = xnn_setup_convolution2d_nhwc_f32(
+      context.convolution_op.get(),                   // operator
+      input_nhwc.size(Layout::Activation4D::Batch),   // batch_size
+      input_nhwc.size(Layout::Activation4D::Height),  // input_height
+      input_nhwc.size(Layout::Activation4D::Width),   // input_width
+      input_nhwc.data_ptr<float>(),                   // input
+      output.data_ptr<float>(),                       // output
+      nullptr);                                       // threadpool
+
+  TORCH_CHECK(
+      xnn_status_success == setup_status,
+      "xnn_setup_convolution2d_nhwc_f32 failed!");
+
+  const xnn_status run_status = xnn_run_operator(
+      context.convolution_op.get(), // operator
+      nullptr);                     // threadpool
+
+  TORCH_INTERNAL_ASSERT(
+      xnn_status_success == run_status,
+      "xnn_run_operator failed!");
+
+  return output.contiguous(input.suggest_memory_format());
+}
+
+Tensor create_and_run(
+    const Tensor& input,
+    const Tensor& weight,
+    const Tensor& bias,
+    const IntArrayRef padding,
+    const IntArrayRef stride,
+    const IntArrayRef dilation,
+    const int64_t groups,
+    const float output_min,
+    const float output_max) {
+  return run(
+      create(
+          weight,
+          bias,
+          padding,
+          stride,
+          dilation,
+          groups,
+          output_min,
+          output_max),
+      input);
+}
+
+} // namespace
+} // namespace convolution2d
+} // namespace internal
+
+bool use_convolution2d(
+    const Tensor& input,
+    const Tensor& weight,
+    const Tensor& bias,
+    const IntArrayRef padding,
+    const IntArrayRef stride,
+    const IntArrayRef dilation,
+    const int64_t groups) {
+  return internal::convolution2d::available(
+            weight,
+            bias,
+            padding,
+            stride,
+            dilation,
+            groups,
+            internal::convolution2d::Context::kMin,
+            internal::convolution2d::Context::kMax) &&
+         internal::convolution2d::usable(input);
+}
+
+Tensor convolution2d(
+    const Tensor& input,
+    const Tensor& weight,
+    const Tensor& bias,
+    const IntArrayRef padding,
+    const IntArrayRef stride,
+    const IntArrayRef dilation,
+    const int64_t groups) {
+  return internal::convolution2d::create_and_run(
+      input,
+      weight,
+      bias,
+      padding,
+      stride,
+      dilation,
+      groups,
+      internal::convolution2d::Context::kMin,
+      internal::convolution2d::Context::kMax);
+}
+
+} // namespace xnnpack
+
+at::Tensor conv2d_prepack(
+    const Tensor& weight,
+    const Tensor& bias,
+    const IntArrayRef stride,
+    const IntArrayRef padding,
+    const IntArrayRef dilation,
+    const int64_t groups,
+    const c10::optional<double> output_min,
+    const c10::optional<double> output_max) {
+  return cpp_custom_type_hack::create(
```

Comment:
cc @jamesr66a torchbind is supposed to replace this, right? Is it ready for this use case?

Reply:
Yes, torchbind is supposed to replace this but using it for PyTorch-internal ops is blocked by working around the aten/libtorch build separation